File: packages/oxlint-plugin-react-doctor/src/plugin/rules/a11y/base-ui-dialog-popup-requires-title.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { findJsxAttribute } from "../../utils/find-jsx-attribute.js";
import { getTrailingJsxNameSegment } from "../../utils/get-trailing-jsx-name-segment.js";
import { hasJsxPropIgnoreCase } from "../../utils/has-jsx-prop-ignore-case.js";
import { hasJsxSpreadAttribute } from "../../utils/has-jsx-spread-attribute.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { resolveNamespacedPartName } from "../../utils/resolve-namespaced-part-name.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { RuleVisitors } from "../../utils/rule-visitors.js";
import { scanJsxSubtreeForPart } from "../../utils/scan-jsx-subtree-for-part.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";

interface BaseUiDialogSurface {
  readonly namespaceName: string;
  // The component subpath and the package root both export the namespace;
  // both package names (pre-1.0 and the 1.0 rename) resolve it.
  readonly moduleSourcePattern: RegExp;
}

// Base UI's Dialog.Popup is the `role="dialog"` element; it takes its
// accessible name from a Dialog.Title (wired automatically) or an explicit
// aria-label. Without either, assistive technology announces an unnamed
// dialog.
const BASE_UI_DIALOG_SURFACES: ReadonlyArray<BaseUiDialogSurface> = [
  {
    namespaceName: "Dialog",
    moduleSourcePattern: /^@base-ui(?:-components)?\/react(?:\/dialog)?$/,
  },
  {
    namespaceName: "AlertDialog",
    moduleSourcePattern: /^@base-ui(?:-components)?\/react(?:\/alert-dialog)?$/,
  },
];

const NAME_PROVIDING_ATTRIBUTES = ["aria-label", "aria-labelledby", "title"] as const;

const renderPropMayProvideName = (attribute: EsTreeNodeOfType<"JSXAttribute">): boolean => {
  if (!attribute.value || !isNodeOfType(attribute.value, "JSXExpressionContainer")) return true;
  const expression = stripParenExpression(attribute.value.expression);
  if (!isNodeOfType(expression, "JSXElement")) return true;
  const renderedAttributes = expression.openingElement.attributes;
  return (
    hasJsxSpreadAttribute(renderedAttributes) ||
    NAME_PROVIDING_ATTRIBUTES.some((name) => hasJsxPropIgnoreCase(renderedAttributes, name))
  );
};

const isTitleElementName = (
  elementName: EsTreeNode,
  surface: BaseUiDialogSurface,
  context: RuleContext,
): boolean => {
  if (
    resolveNamespacedPartName(
      elementName,
      surface.moduleSourcePattern,
      surface.namespaceName,
      context,
    ) === "Title"
  ) {
    return true;
  }
  const trailingSegment = getTrailingJsxNameSegment(elementName);
  return trailingSegment === "Title" || trailingSegment === `${surface.namespaceName}Title`;
};

export const baseUiDialogPopupRequiresTitle = defineRule({
  id: "base-ui-dialog-popup-requires-title",
  title: "Base UI dialog popup without a title",
  severity: "warn",
  requires: ["base-ui"],
  recommendation:
    "Give every Base UI Dialog.Popup and AlertDialog.Popup a Title part (visually hidden when the design shows no heading) or name the dialog with aria-label.",
  create: (context: RuleContext): RuleVisitors => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      for (const surface of BASE_UI_DIALOG_SURFACES) {
        if (
          resolveNamespacedPartName(
            node.name,
            surface.moduleSourcePattern,
            surface.namespaceName,
            context,
          ) !== "Popup"
        ) {
          continue;
        }
        // A spread can supply aria-label; a `render` prop swaps in an
        // element that may carry the name itself.
        if (hasJsxSpreadAttribute(node.attributes)) return;
        const renderAttribute = findJsxAttribute(node.attributes, "render");
        if (renderAttribute && renderPropMayProvideName(renderAttribute)) return;
        if (
          NAME_PROVIDING_ATTRIBUTES.some((attribute) =>
            hasJsxPropIgnoreCase(node.attributes, attribute),
          )
        ) {
          return;
        }
        const element = node.parent;
        if (!element || !isNodeOfType(element, "JSXElement") || element.children.length === 0) {
          return;
        }
        const scan = scanJsxSubtreeForPart(element.children, {
          isPartElementName: (elementName) => isTitleElementName(elementName, surface, context),
          // Same-surface parts (Close, Description, …) are known leaves; any
          // unresolved custom component may render the title itself, so the
          // claim becomes unprovable. Opaque elements still recurse, so a
          // title nested through them counts.
          isOpaqueElement: (childElement) => {
            const childName = childElement.openingElement.name;
            if (
              resolveNamespacedPartName(
                childName,
                surface.moduleSourcePattern,
                surface.namespaceName,
                context,
              ) !== null
            ) {
              return false;
            }
            const trailingSegment = getTrailingJsxNameSegment(childName);
            return (
              trailingSegment !== null &&
              /^[A-Z]/.test(trailingSegment) &&
              trailingSegment !== "Fragment"
            );
          },
        });
        if (scan.foundPart || scan.sawOpaqueContent) return;
        context.report({
          node: node.name,
          message: `This ${surface.namespaceName}.Popup renders no ${surface.namespaceName}.Title, so the dialog has no accessible name and assistive technology announces an unnamed dialog. Add a Title part (visually hidden if the design shows no heading) or an aria-label.`,
        });
        return;
      }
    },
  }),
});

File: native/oxlint/rules/base-ui-dialog-popup-requires-title.rs
use std::cell::Cell;

use oxc_ast::{
    ast::{JSXAttributeItem, JSXAttributeValue},
    AstKind,
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{context::LintContext, rule::Rule, utils::has_jsx_prop_ignore_case, AstNode};

const DIALOG_MESSAGE: &str = "This Dialog.Popup renders no Dialog.Title, so the dialog has no accessible name and assistive technology announces an unnamed dialog. Add a Title part (visually hidden if the design shows no heading) or an aria-label.";
const ALERT_DIALOG_MESSAGE: &str = "This AlertDialog.Popup renders no AlertDialog.Title, so the dialog has no accessible name and assistive technology announces an unnamed dialog. Add a Title part (visually hidden if the design shows no heading) or an aria-label.";
const NAME_PROVIDING_ATTRIBUTES: [&str; 3] = ["aria-label", "aria-labelledby", "title"];

#[derive(Debug, Default, Clone)]
pub struct BaseUiDialogPopupRequiresTitle;

declare_oxc_lint!(
    /// Require accessible names on Base UI dialog popups.
    BaseUiDialogPopupRequiresTitle,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require accessible names on Base UI dialog popups.",
);

impl Rule for BaseUiDialogPopupRequiresTitle {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        if !has_capability_or_unspecified(ctx, "base-ui") {
            return;
        }
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        for surface_name in ["Dialog", "AlertDialog"] {
            if resolve_base_ui_dialog_part_name(&opening_element.name, surface_name, ctx)
                != Some("Popup")
            {
                continue;
            }
            if opening_element
                .attributes
                .iter()
                .any(|attribute| matches!(attribute, JSXAttributeItem::SpreadAttribute(_)))
                || NAME_PROVIDING_ATTRIBUTES.iter().any(|attribute_name| {
                    has_jsx_prop_ignore_case(opening_element, attribute_name).is_some()
                })
                || find_jsx_attribute(opening_element, "render")
                    .is_some_and(render_prop_may_provide_name)
            {
                return;
            }
            let parent = ctx.nodes().parent_node(node.id());
            let AstKind::JSXElement(element) = parent.kind() else {
                return;
            };
            if element.children.is_empty() {
                return;
            }
            let found_title = Cell::new(false);
            let saw_opaque_content = Cell::new(false);
            visit_static_jsx_children(
                &element.children,
                &mut |child_element| {
                    let child_name = &child_element.opening_element.name;
                    if is_base_ui_dialog_title_name(child_name, surface_name, ctx) {
                        found_title.set(true);
                        return false;
                    }
                    if find_jsx_attribute(&child_element.opening_element, "render").is_some_and(
                        |attribute| render_attribute_contains_title(attribute, surface_name, ctx),
                    ) {
                        found_title.set(true);
                    }
                    if resolve_base_ui_dialog_part_name(child_name, surface_name, ctx).is_none()
                        && jsx_element_name_trailing_segment(child_name).is_some_and(|name| {
                            name != "Fragment"
                                && name.as_bytes().first().is_some_and(u8::is_ascii_uppercase)
                        })
                    {
                        saw_opaque_content.set(true);
                    }
                    true
                },
                &mut || saw_opaque_content.set(true),
            );
            if found_title.get() || saw_opaque_content.get() {
                return;
            }
            let message = if surface_name == "Dialog" {
                DIALOG_MESSAGE
            } else {
                ALERT_DIALOG_MESSAGE
            };
            ctx.diagnostic(OxcDiagnostic::warn(message).with_label(opening_element.name.span()));
            return;
        }
    }
}

fn resolve_base_ui_dialog_part_name<'a>(
    element_name: &oxc_ast::ast::JSXElementName<'a>,
    surface_name: &str,
    ctx: &LintContext<'a>,
) -> Option<&'static str> {
    let api_path = resolve_jsx_import_api_path(
        element_name,
        |module_source| match surface_name {
            "Dialog" => matches!(
                module_source,
                "@base-ui/react"
                    | "@base-ui/react/dialog"
                    | "@base-ui-components/react"
                    | "@base-ui-components/react/dialog"
            ),
            "AlertDialog" => matches!(
                module_source,
                "@base-ui/react"
                    | "@base-ui/react/alert-dialog"
                    | "@base-ui-components/react"
                    | "@base-ui-components/react/alert-dialog"
            ),
            _ => false,
        },
        ctx,
    )?;
    let [namespace, part_name] = api_path.as_slice() else {
        return None;
    };
    if namespace != surface_name {
        return None;
    }
    match part_name.as_str() {
        "Popup" => Some("Popup"),
        "Title" => Some("Title"),
        _ => Some("Part"),
    }
}

fn is_base_ui_dialog_title_name<'a>(
    element_name: &oxc_ast::ast::JSXElementName<'a>,
    surface_name: &str,
    ctx: &LintContext<'a>,
) -> bool {
    resolve_base_ui_dialog_part_name(element_name, surface_name, ctx) == Some("Title")
        || jsx_element_name_trailing_segment(element_name).is_some_and(|name| {
            name == "Title"
                || (surface_name == "Dialog" && name == "DialogTitle")
                || (surface_name == "AlertDialog" && name == "AlertDialogTitle")
        })
}

fn render_prop_may_provide_name(attribute: &oxc_ast::ast::JSXAttribute<'_>) -> bool {
    let Some(JSXAttributeValue::ExpressionContainer(container)) = &attribute.value else {
        return true;
    };
    let Some(oxc_ast::ast::Expression::JSXElement(element)) = container
        .expression
        .as_expression()
        .map(oxc_ast::ast::Expression::get_inner_expression)
    else {
        return true;
    };
    element
        .opening_element
        .attributes
        .iter()
        .any(|attribute| matches!(attribute, JSXAttributeItem::SpreadAttribute(_)))
        || NAME_PROVIDING_ATTRIBUTES.iter().any(|attribute_name| {
            has_jsx_prop_ignore_case(&element.opening_element, attribute_name).is_some()
        })
}

fn render_attribute_contains_title<'a>(
    attribute: &oxc_ast::ast::JSXAttribute<'a>,
    surface_name: &str,
    ctx: &LintContext<'a>,
) -> bool {
    let Some(value) = &attribute.value else {
        return false;
    };
    let value_span = value.span();
    ctx.nodes().iter().any(|candidate| {
        let AstKind::JSXOpeningElement(opening_element) = candidate.kind() else {
            return false;
        };
        let candidate_span = opening_element.span;
        candidate_span.start >= value_span.start
            && candidate_span.end <= value_span.end
            && is_base_ui_dialog_title_name(&opening_element.name, surface_name, ctx)
    })
}
